### PR TITLE
ERC4626 Linear Pool

### PR DESCRIPTION
**Pull request by gtaschuk**
This implements a linear pool for a ERC4626 token

Written in collaboration with the Overnight.fi team - This was created to work with https://github.com/ovnstable/ovnstable-app/tree/static-usdplus

This supercedes #1105 

References:
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4626.md

https://ethereum-magicians.org/t/eip-4626-yield-bearing-vault-standard/7900/46

**NOTE: This works for fee-free ERC4626 tokens, and we are working on using `totalAssets/totalSupply` rather than `assetsPerShare` which would incorporate fees, which would better be accounted for in the swap fees**

### CODE REVIEW

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (TomAFrench) [inline pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]:
```suggestion
        uint256 rate = IERC4626(getWrappedToken()).assetsPerShare();
        
        // Upscale to e18
        return rate * _scalingFactor(getWrappedToken());
```

`10 ** decimalsDifference == scalingFactor` so we can avoid making a call out to the wrappedToken here.

We can also pull `_wrappedToken` from `LinearPool` by making `getWrappedToken()` public to avoid having to store it separately here.

Comment (TomAFrench) [inline pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]: Made a comment (now deleted) here before that we can just use the scaling factor from `LinearPool` but that includes a rate can't be used.

We can store `10 ** decimalsDifference` in an immutable variable to avoid having to make many calls out to `_wrappedToken` however.

We can also make `getWrappedToken()` in `LinearPool` public to avoid having to store `_wrappedToken` again in this contract.

Comment (gtaschuk) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]: @mosinnik would you add https://github.com/balancer-labs/balancer-v2-monorepo/commit/4939eafab6d10bf55df9af7c3c15ca581bc01931 to this branch re: storing scaling in an immutable var?

@TomAFrench - I wasn't sure if I should modify `LinearPool` if it's in the wild already.  Should I add the helper method to `LinearPool` or this derived contract?

File: pkg/standalone-utils/contracts/BatchRelayerLibrary.sol
Comment (gtaschuk) [inline pkg/standalone-utils/contracts/BatchRelayerLibrary.sol:31]: We will likely want a `ERC4246Wrapping` lib that wraps/unwraps ERC4246 tokens

UsdPlus implements the methods necessary to be treated as a ERC4246 by the linear pool, but not the full spec

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (TomAFrench) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]: It's fine to modify LinearPool in this way. We won't be changing the interface by making such a change.

File: pkg/standalone-utils/contracts/BatchRelayerLibrary.sol
Comment (rabmarut) [reply pkg/standalone-utils/contracts/BatchRelayerLibrary.sol:31]: Could we add ERC-4626 `deposit` and `redeem` functions to the UsdPlus wrapper? We could simply forward those calls to the existing `wrap` and `unwrap`, and it would go a long way toward generalizing this PR. We would also have to add the `Deposit` and `Withdraw` events.

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (gtaschuk) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]: :+1:  done https://github.com/balancer-labs/balancer-v2-monorepo/commit/85aa152fdd09bc205a017cabc07da15eeee0bfc5

File: pkg/standalone-utils/contracts/BatchRelayerLibrary.sol
Comment (gtaschuk) [reply pkg/standalone-utils/contracts/BatchRelayerLibrary.sol:31]: If @mosinnik is cool with https://github.com/ovnstable/ovnstable-app/pull/13/files than I'm all for it

File: pkg/pool-linear/contracts/interfaces/IERC4626.sol
Comment (mosinnik) [inline pkg/pool-linear/contracts/interfaces/IERC4626.sol:30]: now it have been renamed to `asset()` at EIP4626 draft: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4626.md#asset

File: pkg/standalone-utils/contracts/BatchRelayerLibrary.sol
Comment (mosinnik) [reply pkg/standalone-utils/contracts/BatchRelayerLibrary.sol:31]: it's ok, now I'm working on implementation deposit and redeem methods near current wrap/unwrap

Comment (mosinnik) [reply pkg/standalone-utils/contracts/BatchRelayerLibrary.sol:31]: @gtaschuk new methods implemented in https://github.com/ovnstable/ovnstable-app/pull/10, so we ready to rework wrapper to ERC4626

File: pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol
Comment (TomAFrench) [inline pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol:62]:
```suggestion
    function unwrapERC4626(
```

File: pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol
Comment (TomAFrench) [inline pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol:44]: The name `dynamicToken` is Aave specific language which doesn't apply here.

File: pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol
Comment (TomAFrench) [inline pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol:54]: Stale

File: pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol
Comment (TomAFrench) [inline pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol:80]: Aave specific language.

File: pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol
Comment (TomAFrench) [inline pkg/standalone-utils/contracts/relayer/ERC4626Wrapping.sol:81]:
```suggestion
        uint256 result = wrappedToken.redeem(amount, recipient, address(this));
```

This will either revert or strand user's tokens on the relayer if called with the user as the sender. This is currently pulling tokens from the user onto the relayer and then attempting to redeem tokens which are still on the user's address.

File: pkg/pool-linear/test/ERC4626LinearPool.test.ts
Comment (TomAFrench) [inline pkg/pool-linear/test/ERC4626LinearPool.test.ts:75]: Aave specific language.

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (TomAFrench) [inline pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:65]:
```suggestion
        // Exchange rate between wrapped and underlying token with _wrappedToken.decimals() decimals of precision
        uint256 rate = IERC4626(getWrappedToken()).assetsPerShare();

        // This function returns a 18 decimal fixed point number so upscale to be as if _wrappedToken had 18 decimals
        return rate * _wrappedTokenRateScale;
```

**Review by TomAFrench — CHANGES_REQUESTED** [review]

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (mosinnik) [inline pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:52]: it's wrong check, with this check we'll got pool like dynamicToken<->staticToken, where staticToken is 4626 static wrapper about same dynamicToken. So pool will look like dynToken<->4626(dynToken) and that pull have no sense. 
This check come from aave pool where it check underlying of underlying token.

**Review by rabmarut — DISMISSED** [review]

**Review by rabmarut — COMMENTED** [review]
Looks good from my end, but waiting on final review from @nventuro before merging.

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (mosinnik) [inline pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol:62]: in current version of eip4626 assetsPerShare method has been removed (https://github.com/ethereum/EIPs/pull/4806) so need to be reworked to totalAssets/totalSupply

Comment (mosinnik) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol]: add fix

Comment (gtaschuk) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol]: Awesome.  Thanks for tackling this

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (mosinnik) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol]: done

File: pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol
Comment (mosinnik) [reply pkg/pool-linear/contracts/erc4626/ERC4626LinearPool.sol]: done

**Review by rabmarut — APPROVED** [review]

**Pull request merged**
